### DISCUSSION TITLE
How to achieve single instance app in C#?

### DISCUSSION THREAD

**Discussion by JohnCido** [microsoft/WindowsAppSDK#1747] [Q&A] [answered]
I want my app to have only one instance. I've gone through the [doc](https://docs.microsoft.com/en-us/windows/apps/windows-app-sdk/applifecycle/applifecycle-instancing) and all I found are so many C++ examples and just explanations of differences between single instance and multiple instances app.

Is there a clear example to achieve it using clean C# codes and API interface in 1.0 preview 3?
And why not provide some C# examples in the doc?

**Comment by andrewleader** [comment, marked as answer]
Hi @JohnCido, we're working on improving the docs for single instancing, in the meantime here's a GIST that explains how to accomplish single instance in a C# app using 1.0 Preview 3: **https://gist.github.com/andrewleader/5adc742fe15b06576c1973ea6e999552**

Make sure you build/compile your app for x64, there's currently a known bug on x86.

**Reply by JohnCido** [reply]
I've tried the gist codes and it's not working. I tried to see breakpoint in the `OnLaunched` handler and the `mainInstance.IsCurrent` cannot be evaluated. When I launch the app again, the breakpoint inside `if (!mainInstance.IsCurrent) {...}` was never hit.

**Reply by JohnCido** [reply]
By the way, I was compiling for Debug x64, so I suppose it's not an x86 issue.

**Reply by andrewleader** [reply]
Can you share a full code sample of what you're doing?

**Reply by andrewleader** [reply]
@JohnCido I've published a full code sample here: https://github.com/andrewleader/RedirectActivationWinUI3Sample

You should be able to clone that, set it to `Debug` `x64`, deploy it, and launch the app, then launch a second instance of it and it'll redirect activation and the button on the first instance will also update to say redirected along with coming to the foreground.

**Reply by JohnCido** [reply]
I basically coped the example codes from the gist, with some tweaks of custom variable naming and `FocusAppWindow` helper function:
```C#
/// <summary>
/// Invoked when the application is launched normally by the end user.  Other entry points
/// will be used such as when the application is launched to open a specific file.
/// </summary>
/// <param name="args">Details about the launch request and process.</param>
protected override async void OnLaunched(LaunchActivatedEventArgs args)
{
    AppActivationArguments appArgs = AppInstance.GetCurrent().GetActivatedEventArgs();
    AppInstance mainInstance = AppInstance.FindOrRegisterForKey("main");

    if (!mainInstance.IsCurrent)
    {
        // Redirect to active instance
        await mainInstance.RedirectActivationToAsync(appArgs);
        // Quit the app
        System.Diagnostics.Process.GetCurrentProcess().Kill();
        return;
    }

    AppInstance.GetCurrent().Activated += App_Activated;
    _mainWindow = new MainWindow();
}

private void App_Activated(object sender, AppActivationArguments e)
{
    IntPtr windowHandle = WinRT.Interop.WindowNative.GetWindowHandle(_mainWindow);
    Helpers.FocusAppWindow(windowHandle);
}

private Window _mainWindow;
```
I've checked the codes again and again, everything seems to match. I downloaded your example project and it runs smoothly on my PC.

Could this have anything to do with targeting .NET 6?
When I changed it back to .NET 5 and I can see the correct `mainInstance.IsCurrent` evaluation, but the value is **always** true. So the if statement still gets no hit.

**Reply by KWodarczyk** [reply]
I get issue "The type or namespace name 'Win32' does not exist in the namespace 'Windows'"  is this meant to work in .Net 6 ?

**Reply by andrewleader** [reply]
Hi @KWodarczyk, make sure you completed steps 1 and 2 from the GIST, which add the Win32 methods.

> Install the Microsoft.Windows.CsWin32 NuGet package
> Add the NativeMethods.txt file with two methods, ShowWindow and SetForegroundWindow (the WinUI 3 Activate and AppWindow Show methods don't work)

Alternatively, you can use DllImports to import those two Win32 methods as seen below.

```csharp
        [DllImport("user32.dll")]
        private static extern bool ShowWindow(IntPtr hWnd, int nCmdShow);

        [DllImport("user32.dll")]
        [return: MarshalAs(UnmanagedType.Bool)]
        private static extern bool SetForegroundWindow(IntPtr hWnd);
```

**Reply by KWodarczyk** [reply]
ok that file, NativeMethods.txt what does it do ? i thought it was some sort of read me or something

**Reply by andrewleader** [reply]
The NativeMethods.txt file declares which Win32 methods you want to be projected into C# (so you have to specify the methods you then subsequently want to use). You can read more about it on the CsWin32 docs, they chose not to import all methods by default because it has perf impacts to import every single method, so you have to declare which methods you want.

**Reply by KWodarczyk** [reply]
Ok I can confirm that it works, now :) BTW are there plans to make it less cumbersome ? Like without all that Pinvokes and additional files just pure c# ?

**Reply by andrewleader** [reply]
We do have aspirations to make this easier in the future, however we don't have anything concrete planned to share around that at this time.

**Reply by KWodarczyk** [reply]
Can't we just use Application.Current.Exit(); instead of System.Diagnostics.Process.GetCurrentProcess().Kill(); ?

**Reply by KWodarczyk** [reply]
Also, instead of all that pinvoke stuff i just do   `m_window.DispatcherQueue.TryEnqueue(() => { m_window?.Activate(); });` and works, it this not allowed ?

**Reply by KWodarczyk** [reply]
My OnActivated method looks like this:

`private void OnActivated(object sender, AppActivationArguments e)
        {
            m_window.DispatcherQueue.TryEnqueue(() => { m_window?.Activate(); });

            //// Bring the window to the foreground... first get the window handle...
            //var hwnd = (Windows.Win32.Foundation.HWND)WinRT.Interop.WindowNative.GetWindowHandle(m_window);

            //// Restore window if minimized... requires Microsoft.Windows.CsWin32 NuGet package and a NativeMethods.txt file with ShowWindow method
            //Windows.Win32.PInvoke.ShowWindow(hwnd, Windows.Win32.UI.WindowsAndMessaging.SHOW_WINDOW_CMD.SW_RESTORE);

            //// And call SetForegroundWindow... requires Microsoft.Windows.CsWin32 NuGet package and a NativeMethods.txt file with SetForegroundWindow method
            //Windows.Win32.PInvoke.SetForegroundWindow(hwnd);
        }`

**Reply by hintsofttech** [reply]
**single instance** should mean **single instance per user**
because multiple users may run the app through remote desktop connections...
